_posts/2018-06-01-Automated-Canary-Analysis-using-Spinnaker.md

```diff
@@ -0,0 +1,871 @@
+
+```
+hal config provider docker-registry account add gcr-eu \
+    --address eu.gcr.io \
+    --password-file ~/.gcp/gcp.json \
+    --username _json_key
+
+hal config provider kubernetes account edit my-k8s-account --docker-registries my-gcr-account gcr-eu
+```
+
+#### IAM Configuration
+
+Enable Stackdriver access for Spinnaker in [GCP's IAM settings](https://console.cloud.google.com/iam-admin/iam).
+
+Add the following roles to the member with name `gcs-service-account`:
+- Logging Admin
+- Monitoring Admin
+
+#### Automated Canary Analysis
+
+Before you perform hal deploy apply, enable automated canary analysis.
+Follow the guide further down, but first of all set some variables while still SSH’d in the Halyard VM.
+One of these variables is the Spinnaker bucket automatically created when installing Halyard.
+Look for the right bucket identifier in the GCP GKE buckets dashboard.
+
+```
+PROJECT_ID=$(gcloud info --format='value(config.project)')
+JSON_PATH=~/.gcp/gcp.json
+MY_SPINNAKER_BUCKET=spin-48b89b5e-dd67-446a-ad9f-66e8783e9822
+```
+
+Follow the [official canary quickstart documentation](https://www.spinnaker.io/setup/canary/#quick-start).
+
+Configure the default metrics store.
+```
+hal config canary edit --default-metrics-store stackdriver
+```
+
+And finally execute the rollout.
+
+```
+hal deploy apply
+```
+
+### Troubleshooting
+
+#### Authentication
+Sometimes an issue might occur with credentials on the Halyard VM:
+
+```
+! ERROR Unable to communicate with your Kubernetes cluster: Failure
+  executing: GET at: https://35.205.113.166/api/v1/namespaces. Message: Forbidden!
+  User gke_spinnaker-demo-184310_europe-west1-d_spinnaker-alpha doesn't have
+  permission. namespaces is forbidden: User "client" cannot list namespaces at the
+  cluster scope: Unknown user "client"..
+? Unable to authenticate with your Kubernetes cluster. Try using
+  kubectl to verify your credentials.
+```
+
+In this case, enable legacy authentication in the GKE UI for the cluster.
+
+#### Cluster Debugging
+You can monitor deployment locally on your own PC by running `kubectl get pods -w --all-namespaces`.
+For this to work, kubectl needs permissions to talk to the cluster.
+You can use gcloud to populate your kubectl config file with credentials to access the cluster.
+This can help you to look into specific logs of each Spinnaker pod or follow up on deployments handled by Spinnaker.
+
+#### Audit Logging
+
+You can find out which commands are sent to GCP by enabling audit logging.
+Turn on audit logging: https://cloud.google.com/monitoring/audit-logging & https://cloud.google.com/logging/docs/audit/configure-data-access#example
+
+<a name="commands" />
+
+### Comprehensive list of commands
+
+These are all the commands we executed in order to get everything set up.
+Fill in the `<PLACEHOLDER>` placeholders according to your preferences.
```

Comment:
It might help to raise the placeholders to the top like

```
ZONE=<ZONE>
PROJECT_NAME=<PROJECT_NAME>
```

and reference them later as bash variables `$ZONE`, `$PROJECT_NAME`...